File: lista python/at4.py
class Animal:
    def __init__(self, nome):
        self.nome = nome
    
    def som (self):
        raise NotImplementedError("Subclasses devem implementar o método som.")
    

class Cachorro(Animal):
    def som(self):
        return f"{self.nome}: Au au!"
    

class Gato(Animal):
     def som(self):
        return f"{self.nome}: Miau!"
     

cachorro = Cachorro("Mike")
gato = Gato("Garfield")

print(cachorro.som())
print(gato.som())

File: lista Go/at4.go
package main

import "fmt"


type Animal interface {
    Som()  
}


type Cachorro struct{}


func (c Cachorro) Som() {
    fmt.Println("O cachorro late: Au Au!")
}


type Gato struct{}


func (g Gato) Som() {
    fmt.Println("O gato mia: Miau!")
}

func main() {
    
    var animal Animal

    animal = Cachorro{}
    animal.Som()  

    animal = Gato{}
    animal.Som()  
}
